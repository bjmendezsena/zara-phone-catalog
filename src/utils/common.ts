export const isClientSide = (): boolean => {
  return typeof window !== "undefined";
};

export const getValue = <T extends Record<string, unknown>>(
  source: T,
  key: string
): unknown => {
  const arrKeys = key.split(".") as string[];
  const firstKey = arrKeys.shift() || "";
  const newSource = source[firstKey as keyof typeof source];

  if (arrKeys.length > 0 && newSource && typeof newSource === "object") {
    return getValue(newSource as Record<string, unknown>, arrKeys.join("."));
  }

  return newSource;
};
