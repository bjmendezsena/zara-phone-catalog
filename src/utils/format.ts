export const formatCurrency = (amount: number): string => {
  return `${amount.toFixed(2)} EUR`;
};

export function formatString(fmt: string, ...args: any[]): string {
  const re = /(%?)(%([ojds]))/g;

  if (args.length) {
    fmt = fmt.replace(re, (match, escaped, _ptn, flag) => {
      let arg = args.shift();

      switch (flag) {
        case "o":
          if (Array.isArray(arg)) {
            arg = JSON.stringify(arg);
          }
          break;
        case "s":
          arg = "" + arg;
          break;
        case "d":
          arg = Number(arg);
          break;
        case "j":
          arg = JSON.stringify(arg);
          break;
        default:
          break;
      }

      if (!escaped) {
        return arg;
      }

      args.unshift(arg);
      return match;
    });
  }

  // arguments remain after formatting
  if (args.length) {
    fmt += " " + args.join(" ");
  }

  // update escaped %% values
  fmt = fmt.replace(/%{2,2}/g, "%");

  return "" + fmt;
}
