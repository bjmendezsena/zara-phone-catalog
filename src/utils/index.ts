export * from "./cn";
export * from "./uuid";
export * from "./create-context";
export * from "./json";
export * from "./common";
export * from "./format";
