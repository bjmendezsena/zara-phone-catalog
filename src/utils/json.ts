export const isValidJSON = (
  value: unknown
): value is Record<string, unknown> => {
  if (typeof value !== "string") return false;
  try {
    JSON.parse(value);
    return true;
  } catch {
    return false;
  }
};
