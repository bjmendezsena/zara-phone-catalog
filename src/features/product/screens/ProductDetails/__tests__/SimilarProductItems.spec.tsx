import { render, getHtml, mockProductDetailsResponse } from "@/tests";
import {
  SimilarProductItems,
  SimilarProductItemsProps,
} from "@/features/product/screens/ProductDetails/SimilarProductItems";
import { formatCurrency } from "@/utils";

describe(`<${SimilarProductItems.name}/>`, () => {
  const defaultProps: SimilarProductItemsProps = {
    products: mockProductDetailsResponse.similarProducts,
  };

  it("renders without crashing", () => {
    const { container } = render(<SimilarProductItems {...defaultProps} />);
    expect(getHtml(container)).toMatchSnapshot();
  });

  it("displays the correct number of similar products", () => {
    const { getAllByTestId } = render(
      <SimilarProductItems {...defaultProps} />
    );
    const productCards = getAllByTestId("product-card");
    expect(productCards).toHaveLength(defaultProps.products.length);
  });

  it("renders each product card with correct details", () => {
    const { getAllByAltText, getByText, findAllByText } = render(
      <SimilarProductItems {...defaultProps} />
    );

    defaultProps.products.forEach((product) => {
      expect(getAllByAltText(`${product.brand} ${product.name}`)).toBeTruthy();
      expect(getByText(product.name)).toBeTruthy();
      expect(findAllByText(product.brand)).toBeTruthy();
      expect(findAllByText(formatCurrency(product.basePrice))).toBeTruthy();
    });
  });
});
