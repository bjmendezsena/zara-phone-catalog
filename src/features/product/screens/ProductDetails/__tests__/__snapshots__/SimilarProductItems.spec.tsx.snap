// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<SimilarProductItems/> renders without crashing 1`] = `
"<div>
  <div
    class="mt-16 xl:pl-[360px]"
  >
    <h2
      class="text-xl font-light text-primary mb-8 uppercase tracking-wide pl-4 xl:pl-0"
    >
      SIMILAR ITEMS
    </h2>
    <div
      class="relative overflow-hidden"
    >
      <div
        class="flex  overflow-x-auto scrollbar-thin scrollbar-track-transparent scrollbar-thumb-primary pb-4 px-4 xl:pl-0 xl:pr-6"
        style="scrollbar-width: thin; scrollbar-color: var(--color-primary) transparent;"
      >
        <div
          class="flex-shrink-0 w-[280px] sm:w-[320px] md:w-[240px] lg:w-[280px] xl:w-[320px]"
        >
          <a
            aria-label="Ver detalles del producto Xiaomi Redmi 12"
            class="w-full h-full max-h-[344px] flex items-center justify-center text-center transition-all duration-300 focus:outline-none focus:ring-2 focus:ring-black focus:ring-offset-2 border-1 flex-col relative overflow-hidden group"
            data-testid="product-card"
            href="/product/XIA-R12"
          >
            <div
              class="absolute inset-0 bg-primary transform translate-y-full group-hover:translate-y-0 transition-transform duration-500 ease-out"
            />
            <div
              class="relative z-10 w-full h-full flex flex-col"
            >
              <div
                aria-label="Imagen del producto Xiaomi Redmi 12"
                class="flex-1 flex items-center justify-center p-4"
              >
                <div
                  class="w-full max-w-[150px] h-full max-h-[200px] flex items-center justify-center"
                >
                  <img
                    alt="Xiaomi Redmi 12"
                    class="w-full h-full object-contain group-hover:brightness-110 transition-all duration-300"
                    data-priority="true"
                    height="200"
                    src="http://prueba-tecnica-api-tienda-moviles.onrender.com/images/XIA-R12-sky-blue.webp"
                    width="200"
                  />
                </div>
              </div>
              <div
                class="flex-shrink-0 w-full flex justify-between items-end px-4 py-3"
              >
                <div
                  class="flex flex-col space-y-1 items-start"
                >
                  <span
                    class="text-xs text-secondary font-light group-hover:text-white transition-colors duration-300 uppercase tracking-wide"
                  >
                    Xiaomi
                  </span>
                  <span
                    class="text-sm text-primary font-light group-hover:text-white transition-colors duration-300 leading-tight"
                  >
                    Redmi 12
                  </span>
                </div>
                <span
                  class="text-sm text-primary font-medium group-hover:text-white transition-colors duration-300"
                >
                  117.29 EUR
                </span>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex-shrink-0 w-[280px] sm:w-[320px] md:w-[240px] lg:w-[280px] xl:w-[320px]"
        >
          <a
            aria-label="Ver detalles del producto Apple iPhone 15 Pro Max"
            class="w-full h-full max-h-[344px] flex items-center justify-center text-center transition-all duration-300 focus:outline-none focus:ring-2 focus:ring-black focus:ring-offset-2 border-1 flex-col relative overflow-hidden group"
            data-testid="product-card"
            href="/product/APL-I15PM"
          >
            <div
              class="absolute inset-0 bg-primary transform translate-y-full group-hover:translate-y-0 transition-transform duration-500 ease-out"
            />
            <div
              class="relative z-10 w-full h-full flex flex-col"
            >
              <div
                aria-label="Imagen del producto Apple iPhone 15 Pro Max"
                class="flex-1 flex items-center justify-center p-4"
              >
                <div
                  class="w-full max-w-[150px] h-full max-h-[200px] flex items-center justify-center"
                >
                  <img
                    alt="Apple iPhone 15 Pro Max"
                    class="w-full h-full object-contain group-hover:brightness-110 transition-all duration-300"
                    data-priority="true"
                    height="200"
                    src="http://prueba-tecnica-api-tienda-moviles.onrender.com/images/APL-I15PM-titanio-negro.webp"
                    width="200"
                  />
                </div>
              </div>
              <div
                class="flex-shrink-0 w-full flex justify-between items-end px-4 py-3"
              >
                <div
                  class="flex flex-col space-y-1 items-start"
                >
                  <span
                    class="text-xs text-secondary font-light group-hover:text-white transition-colors duration-300 uppercase tracking-wide"
                  >
                    Apple
                  </span>
                  <span
                    class="text-sm text-primary font-light group-hover:text-white transition-colors duration-300 leading-tight"
                  >
                    iPhone 15 Pro Max
                  </span>
                </div>
                <span
                  class="text-sm text-primary font-medium group-hover:text-white transition-colors duration-300"
                >
                  1319.00 EUR
                </span>
              </div>
            </div>
          </a>
        </div>
        <div
          class="flex-shrink-0 w-[280px] sm:w-[320px] md:w-[240px] lg:w-[280px] xl:w-[320px]"
        >
          <a
            aria-label="Ver detalles del producto OPPO Reno 11 F"
            class="w-full h-full max-h-[344px] flex items-center justify-center text-center transition-all duration-300 focus:outline-none focus:ring-2 focus:ring-black focus:ring-offset-2 border-1 flex-col relative overflow-hidden group"
            data-testid="product-card"
            href="/product/OPP-R11F"
          >
            <div
              class="absolute inset-0 bg-primary transform translate-y-full group-hover:translate-y-0 transition-transform duration-500 ease-out"
            />
            <div
              class="relative z-10 w-full h-full flex flex-col"
            >
              <div
                aria-label="Imagen del producto OPPO Reno 11 F"
                class="flex-1 flex items-center justify-center p-4"
              >
                <div
                  class="w-full max-w-[150px] h-full max-h-[200px] flex items-center justify-center"
                >
                  <img
                    alt="OPPO Reno 11 F"
                    class="w-full h-full object-contain group-hover:brightness-110 transition-all duration-300"
                    data-priority="true"
                    height="200"
                    src="http://prueba..."
`;
